tests/functions/test_user_defined_func.py

```diff
@@ -28,6 +28,294 @@ def myFunction(variable, param1, param2):
+        else:
+            e = U
+        val = benchmark(e, [[1, 3]])
+        assert np.allclose(val, [[5, 9]])
+        val2 = e([[-1, 3]])
+        assert np.allclose(val2, [[5, -3]])
+
+    @pytest.mark.parametrize("bin_execute", ['Python',
+                                             pytest.param('LLVM', marks=pytest.mark.llvm),
+                                             #pytest.param('PTX', marks=[pytest.mark.llvm, pytest.mark.cuda]),
+                                            ])
+    @pytest.mark.benchmark(group="Function UDF")
+    def test_user_def_func_variable_index(self, bin_execute, benchmark):
+        def myFunction(variable):
+            variable[0][0] = variable[0][0] + 5
+            variable[0][1] = variable[0][1] + 7
+            return variable
+
+        U = UserDefinedFunction(custom_function=myFunction, default_variable=[[0, 0]])
+        if (bin_execute == 'LLVM'):
+            e = pnlvm.execution.FuncExecution(U).execute
+        else:
+            e = U
+        val = benchmark(e, [[1, 3]])
+        assert np.allclose(val, [[6, 10]])
+
+    @pytest.mark.parametrize("variable, param1, param2", [
+                        (1, 2, 3), # this is vec - scalar, since udfs automatically convert scalar inputs to vec
+                        (np.ones((2, 2)), 2, 3), # mat - scalar
+                        (np.ones((2)), np.array([1, 2]), np.array([3, 4])), # vec - vec
+                        (np.ones((2, 2)), np.array([np.array([1, 2]), np.array([3, 4])]), np.array([np.array([5, 6]), np.array([7, 8])])) # mat - mat
+                        ])
+    @pytest.mark.parametrize("bin_execute", ['Python',
+                                             pytest.param('LLVM', marks=pytest.mark.llvm),
+                                             #pytest.param('PTX', marks=[pytest.mark.llvm, pytest.mark.cuda]),
+                                            ])
+    @pytest.mark.benchmark(group="Function UDF")
+    def test_user_def_func_add(self, variable, param1, param2, bin_execute, benchmark):
+        def myFunction(variable, param1, param2):
+            return variable + param1 + param2
+
+        U = UserDefinedFunction(custom_function=myFunction, default_variable=variable, param1=param1, param2=param2)
+        if (bin_execute == 'LLVM'):
+            e = pnlvm.execution.FuncExecution(U).execute
+        else:
+            e = U
+        val = benchmark(e, variable)
+        assert np.allclose(val, variable + param1 + param2)
+
+    @pytest.mark.parametrize("variable, param1, param2", [
+                        (1, 2, 3), # this is vec - scalar, since udfs automatically convert scalar inputs to vec
+                        (np.ones((2, 2)), 2, 3), # mat - scalar
+                        (np.ones((2)), np.array([1, 2]), np.array([3, 4])), # vec - vec
+                        (np.ones((2, 2)), np.array([np.array([1, 2]), np.array([3, 4])]), np.array([np.array([5, 6]), np.array([7, 8])])) # mat - mat
```

Comment:
`np.array([[1, 2], [3, 4]]),` is much more readable

Reply:
this doesn't seem to work in pytest (but works in python interpreter). Is there any restriction in pytest?